src/lib.rs

```diff
@@ -0,0 +1,184 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::type_complexity)]
+
+pub use pallet::*;
+
+pub mod types;
+
+pub use types::*;
+
+#[cfg(test)]
+mod tests;
+
+pub use frame_support::{
+	storage::IterableStorageMap,
+	traits::tokens::{
+		currency::{Currency, ReservableCurrency},
+		ExistenceRequirement,
+	},
+	ReversibleStorageHasher,
+};
+
+use sp_runtime::traits::Saturating;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::{pallet_prelude::*, sp_runtime::traits::Zero};
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// The trait to manage funds
+		type Currency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
+		/// The maximum amount of metadata
+		#[pallet::constant]
+		type MaxMetadataLength: Get<u32>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn plan_nonce)]
+	pub type PlanNonce<T: Config> = StorageValue<_, PlanId, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn subscription_plans)]
+	pub type Plans<T: Config> = StorageMap<
+		_,
+		Blake2_256,
```

Comment:
When key is not chose by user, you can use lighter `twox_64_concat` hashing algorithm instead
https://www.shawntabrizi.com/substrate/transparent-keys-in-substrate/